web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
+
+                enddir++;
+                w->command.body = enddir;
+                w->command.length = (size_t) (endcmd - enddir);
+            }
+        }
+    }
+    else{
+        w->directory.body = begin;
+        w->directory.length = w->path.length - 1;
+        w->version.body = NULL;
+        w->version.length = 0;
+        w->command.body = NULL;
+        w->command.length = 0;
+    }
+}
+
+/**
+ *  Set without query string
+ *
+ *  When I do not have query string, it is necessary to set some default values.
+ *
+ * @param w is the structure with all information of the client request.
+ */
+static inline void web_client_set_without_query_string(struct web_client *w) {
+    w->query_string.body = NULL;
+    w->query_string.length = 0;
+
+    char *test = w->path.body+1;
+    if (!strncmp(test,"api/v1/",7) ) {
+        test += 7;
+        char *endofcommand=strchr(test,' ');
+        w->command.length=(endofcommand)?(size_t)endofcommand-(size_t)test:strlen(test);
+    } else {
+        w->command.length = w->path.length;
+    }
+    w->command.body = test;
+    w->total_params = 0;
+}
+
+/**
+ *  Split path and query
+ *
+ *  This function splits path and query string case there is one.
+ *
+ * @param w is the structure with all information of the client request.
+ */
+static inline void web_client_split_path_query(struct web_client *w) {
+    //I am working now with the decoded URL instead the original
+    w->path.body = w->decoded_url;
+    w->decoded_length = strlen(w->decoded_url);
+
+    //Is there a separator?
+    char *moveme = strchr(w->path.body,'?');
+    char *enddir;
+    if (moveme) { //I have a query string
+        w->path.length = moveme - w->path.body;
+        w->query_string.body = moveme;
+        w->query_string.length = w->decoded_length - w->path.length;
+
+        enddir = strchr(w->path.body+1,'/');
+        char *begin = w->path.body+1;
+        web_client_set_directory(w,begin,enddir,moveme);
+        if (w->query_string.body) {
+            enddir = strchr(moveme,'=');
+            if (!web_client_parse_request(w,enddir) ) {
+                moveme++;
+                size_t length = strlen(moveme);
+                w->param_name[0].body = moveme;
+                w->param_name[0].length = length;
+                w->param_values[0].body = moveme;
+                w->param_values[0].length = length;
+
+                w->total_params = 1;
+            }
+        }
+    } else { //I do not have query string
+        w->path.length = w->decoded_length;
+
+        //The fact that I do not have query string, does not mean I do not have directories,
+        //so it is necessary to continue working
+        enddir = strchr(w->path.body+1,'/');
+        w->directory.body = w->path.body + 1;
+        if(enddir) {
+            w->directory.length = (size_t)(enddir - w->directory.body);
+            enddir++;
+
+            w->version.body = enddir;
+            enddir = strchr(++enddir,'/');
```

Comment:
Logic: at this point we've already done `++enddir` twice, can we get out of bounds?

Reply:
It is the same previous explanation about the first line of the request, so we will have at least more 13 bytes in front.